pyproject.toml

```diff
@@ -25,60 +25,67 @@ classifiers = [
     "Programming Language :: Python :: 3.11",
 ]
 
-[build-system]
-requires = [
-    "poetry-core>=1.0.0",
-    "poetry-dynamic-versioning",
-]
-build-backend = "poetry_dynamic_versioning.backend"
-
-[tool.poetry-dynamic-versioning]
-enable = true
-
 [tool.poetry.dependencies]
-python = ">=3.8, <4.0"
+python = ">=3.8, <3.12"
+beautifulsoup4 = "*"
+dataretrieval = "*"
 erddapy = "*"
+geopandas = "*"
+html5lib = "*"
+limits = "*"
+lxml = "*"
+numpy = "*"
 pandas = "*"
 pydantic = "*"
 requests = "*"
-beautifulsoup4 = "*"
-geopandas = "*"
-xarray = "*"
 Shapely = "*"
-limits = "*"
 tqdm = "*"
-html5lib = "*"
-numpy = "*"
-lxml = "*"
-dataretrieval = "*"
+xarray = "*"
 
-[tool.poetry.dev-dependencies]
-mypy = ">=0.991"
-pre-commit = ">=2.0"
-prospector = "<1.8"   # due to https://github.com/PyCQA/prospector/issues/557
-pytest = ">=6.0"
-pytest-xdist = "*"
+[tool.poetry.group.dev.dependencies]
+covdefaults = "*"
+coverage = {version = "*", extras = ["toml"]}
+mypy = ">=1"
+pytest = "*"
 pytest-cov = ">=3.0"
 pytest-recording = "*"
+pytest-xdist = "*"
 types-requests = "*"
```

Comment:
@pmav99 shouldn't `ruff` be on this list?

Reply:
Indeed, we could add it as primary dev dependency, too. Nevertheless, since we also use pre-commit we can also omit it.

For each one of its checks, pre-commit creates a dedicated environment. This means that the checks that are written in python will be installed in their own dedicated virtualenv which on linux systems should exist at `~/.cache/pre-commit/repo*`.

So, since we already have ruff somewhere in our system, why install it in our searvey virtualenv, too? The benefits of omitting it are:

1. we have fewer dependencies. Fewer dependencies means faster dependency resolution and lower chances for dependency conflicts.
2. We have a single source of truth. If we use a tool both via pre-commit and install it in our virtualenv then we must make sure that we keep the versions in sync. For example if a new version of black formats the code differently than the previous one, pre-commit will be complaining that the code is not formatted correctly even though we just run `black ./` and git added the changes. This is something that can lead to some serious head scratching until you realize what the real issue is.

Now, arguably running `pre-commit run ruff -a` is annoying. But that's where the Makefile comes into play. It defines easy to use/discover aliases that alleviate the need of remembering the exact syntax

Reply:
I didn't realize precommit creates env with all the specified test packages automatically. I thought we need to specify it as a main package dependency for the precommit to work. Thanks for clarifying

Reply:
Yes it does. You just need to pay some attention, because some of the tools require to be installed in the same environment as the tested code  (e.g. `pylint` and `mypy`). These tools should be run like this:
https://github.com/oceanmodeling/searvey/blob/50f027c4911313f51377760344543c2e2e0a7260/.pre-commit-config.yaml#L72-L79

Do notice the `language` which is `"system"`. This means that pre-commit should use the locally installed version instead of installing the tool in `~/.cache/pre-commit/repo*`. At the same time, if you are using the https://pre-commit.ci service, you should make sure to skip them
https://github.com/oceanmodeling/searvey/blob/50f027c4911313f51377760344543c2e2e0a7260/.pre-commit-config.yaml#L2-L3

This is the reason, I added `mypy` on `pyproject.toml` and run it on github-actions:
https://github.com/oceanmodeling/searvey/blob/50f027c4911313f51377760344543c2e2e0a7260/.github/workflows/run_tests.yml#L45

For more info check here: https://pre-commit.com/#repository-local-hooks

For the record, it is possibly to ask pre-commit  to install [additional dependencies](https://pre-commit.com/#config-additional_dependencies) in each of its "repos", but I think that you need to specify a list of packages in `.pre-commit.config.yml` and I don't want to have to keep it in sync with the list of dependencies in `pyproject.toml`.

Reply:
Thank you for the detailed explanation.